### PR TITLE
Fix loading of OrderedDict with LibTorch

### LINKED ISSUE #100741: Reading `state_dict()` in C++
### 🐛 Describe the bug

Hi, 

Using `torch::pickle_load` to read a state_dict() serialized in Python fails with:

```
  isGenericDict() INTERNAL ASSERT FAILED at "/Users/dfalbel/Documents/actions-runner/mlverse-m1/_work/libtorch-mac-m1/libtorch-mac-m1/pytorch/aten/src/ATen/core/ivalue_inl.h":1945, please report a bug to PyTorch. Expected GenericDict but got None
Exception raised from toGenericDict at /Users/dfalbel/Documents/actions-runner/mlverse-m1/_work/libtorch-mac-m1/libtorch-mac-m1/pytorch/aten/src/ATen/core/ivalue_inl.h:1945 (most recent call first):
frame #0: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 92 (0x10d39952c in libc10.dylib)
frame #1: c10::detail::torchInternalAssertFail(char const*, char const*, unsigned int, char const*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 72 (0x10d399768 in libc10.dylib)
frame #2: torch::jit::Unpickler::readInstruction() + 12824 (0x294f31cf0 in libtorch_cpu.dylib)
frame #3: torch:
```

AFAICT this is known see  #67902, #37213, #36577 and #95362
`model.state_dict()` returns an `OrderedDict` and Libtorch's bundled unpickler doesn't support reading from it.
Calling `dict(model.state_dict())` before saving is a reasonable workaround, but sometimes user's don't have access to the code that serializes the model.

You can reproduce with eg:

```
std::ifstream file("/Users/dfalbel/Downloads/ordered_dict.pt", std::ios::binary);
  std::vector<char> data((std::istreambuf_iterator<char>(file)),
                         std::istreambuf_iterator<char>());
  auto ivalue = torch::pickle_load(data);
```

Where ordered_dict.pt, is a file saved with:

```
import torch
torch.save(torch.nn.Linear(10,10).state_dict(), "ordered_dict.pt")
```

### Versions

Current main LibTorch

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### PR DESCRIPTION
**Pull request by dfalbel**
Fixes #100741
This PR re-adds changes from #100743 that was reverted, trying to fix the test case so it's more portable.

### CODE REVIEW

**Comment by Skylion007** [comment]
@jeanschmidt since you reverted the OG PR, can you take a look?

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
Imported to check whether it will work, but it probably will require some minor tweaks in internal build system (where all tests are run in isolation to avoid side-effects)

File: test/cpp/jit/test_save_load.cpp [hunk 1]
```diff
@@ -152,6 +152,39 @@ TEST(SerializationTest, TypeTags) {
...
+  // Refer: SaveStateDict in test/cpp/jit/tests_setup.py
+  std::ifstream file("state_dict.pt", std::ios::binary);
+  // a. You need this line, otherwise istream_iterator will skip all the
+  // whitespace in the file
+  //     which will skip characters when reading "ivalue.pt"
+  // b. try std::istreambuf_iterator instead, it seems to me that it doesn't
+  // skip any characters.
+  // see:
+  // https://discuss.pytorch.org/t/reading-tensors-in-c-saved-with-torch-save/93130/2?u=dfalbel
+  file >> std::noskipws;
```
Comment (malfet) [inline test/cpp/jit/test_save_load.cpp:166]: I can be wrong, but this flag have no effect on binary streams.

**Comment by dfalbel** [comment]
Thank you, let me know if there's something I can work on.

**Comment by dfalbel** [comment]
Hi, perhaps should I remove the test that fails internally to make it easier to get this merged?

**Comment by kit1980** [comment]
@malfet how should we proceed with this?

**Comment by dfalbel** [comment]
Hi, I still would love to see this merged into PyTorch as it allows reading Python saved state dicts from C++ without any previous modifications. I understand that the we might need to update the PR. Thanks very much!

**Comment by richardkmichael** [comment]
This is still a problem.  Re-opening and fixing it would be helpful.

**Pull request closed without merging**